utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -53,169 +72,283 @@ pub struct TryRuntimeCmd {
+	/// node with `--rpc-cors=all`
+	/// https://github.com/paritytech/jsonrpsee/issues/346
 	Live {
-		/// An optional state snapshot file to WRITE to. Not written if set to `None`.
+		/// An optional state snapshot file to WRITE to.
 		#[structopt(short, long)]
 		snapshot_path: Option<PathBuf>,
 
 		/// The block hash at which to connect.
 		/// Will be latest finalized head if not provided.
-		#[structopt(short, long, multiple = false, parse(try_from_str = parse_hash))]
+		#[structopt(short, long, multiple = false, parse(try_from_str = parse::hash))]
 		block_at: Option<String>,
 
 		/// The modules to scrape. If empty, entire chain state will be scraped.
 		#[structopt(short, long, require_delimiter = true)]
 		modules: Option<Vec<String>>,
 
 		/// The url to connect to.
-		#[structopt(default_value = "ws://localhost:9944", parse(try_from_str = parse_url))]
+		#[structopt(default_value = "ws://localhost:9944", parse(try_from_str = parse::url))]
 		url: String,
 	},
 }
 
-fn parse_hash(block_number: &str) -> Result<String, String> {
-	let block_number = if block_number.starts_with("0x") {
-		&block_number[2..]
-	} else {
-		block_number
+async fn on_runtime_upgrade<B, ExecDispatch>(shared: SharedParams, command: OnRuntimeUpgradeCmd, config: Configuration) -> sc_cli::Result<()>
+where
+	B: BlockT,
+	B::Hash: FromStr,
+	<B::Hash as FromStr>::Err: Debug,
+	NumberFor<B>: FromStr,
+	<NumberFor<B> as FromStr>::Err: Debug,
+	ExecDispatch: NativeExecutionDispatch + 'static,
+{
+	let spec = config.chain_spec;
+	let genesis_storage = spec.build_storage()?;
+
+	let code = StorageData(
+		genesis_storage
+			.top
+			.get(well_known_keys::CODE)
+			.expect("code key must exist in genesis storage; qed")
+			.to_vec(),
+	);
+	let code_key = StorageKey(well_known_keys::CODE.to_vec());
+
+	let wasm_method = shared.wasm_method;
+	let execution = shared.execution;
+
+	let mut changes = Default::default();
+	// don't really care about these -- use the default values.
+	let max_runtime_instances = config.max_runtime_instances;
+	let heap_pages = config.default_heap_pages;
+	let executor = NativeExecutor::<ExecDispatch>::new(
+		wasm_method.into(),
+		heap_pages,
+		max_runtime_instances,
+	);
+
+	let ext = {
+		let builder = match command.state {
+			State::Snap { snapshot_path } => {
+				Builder::<B>::new().mode(Mode::Offline(OfflineConfig {
+					state_snapshot: SnapshotConfig::new(snapshot_path),
+				}))
+			},
+			State::Live {
+				url,
+				snapshot_path,
+				block_at,
+				modules
+			} => Builder::<B>::new().mode(Mode::Online(OnlineConfig {
+				transport: url.to_owned().into(),
+				state_snapshot: snapshot_path.as_ref().map(SnapshotConfig::new),
+				modules: modules.to_owned().unwrap_or_default(),
+				at: block_at.as_ref()
+					.map(|b| b.parse().map_err(|e| format!("Could not parse hash: {:?}", e))).transpose()?,
+				..Default::default()
+			})),
+		};
+
+		// inject the code into this ext.
+		builder.inject(&[(code_key, code)]).build().await?
```

Comment:
I forgot to add this to the new command, I belive. And without this, we use the code which is on-chain. We need to replace it with the code that comes from the genesis of the current runtime, which basically means "whatever is in the current repository".

Reply:
well, although I am not sure if we want code override for this command or not 🤔

Reply:
Added this in for the offchain worker command; easy to take out if we want